### PR TITLE
Edit Product > Shipping: extract view data logic from view controller to view model

### LINKED ISSUE #1985: Edit Products: extract view data logic from view controller to a separate view model
With the introduction of discard changes action sheet, the logic becomes more complicated especially when fixing price and inventory settings in https://github.com/woocommerce/woocommerce-ios/pull/1979. I intended to add unit tests in that PR, but it wasn't suitable for a release hot fix. Let's spend a little time extracting the view data logic from the view controller to a separate view model for the following screens just to start:

- [x] `ProductPriceSettingsViewController`
- [x] `ProductFormViewController`: thread https://github.com/woocommerce/woocommerce-ios/pull/1977#discussion_r393014680
- [x] `ProductInventorySettingsViewController` https://github.com/woocommerce/woocommerce-ios/pull/2602
- [ ] `ProductShippingSettingsViewController` https://github.com/woocommerce/woocommerce-ios/pull/2714

### LINKED ISSUE #2712: Edit Product > Shipping Settings: tapping "Done" before a shipping class is synced removes the shipping class
**Describe the bug**

In shipping settings, we sync a product's shipping class (if previously set) to ensure the shipping class is up to date for the name and potential edit in the list selector. However, if the user taps "Done" before the shipping class is synced, it removes the shipping class from the product.

**To Reproduce**

- Go to the Products tab
- Tap on a physical product that has a shipping class set
- Tap on the shipping row and quickly tap "Done" before the shipping class row shows the name of the shipping class --> notice the shipping class is removed from the product

**Screenshots**

simulator

**Expected behavior**

The shipping class should not be removed after just tapping "Done" in shipping settings.

**Isolating the problem (mark completed items with an [x]):**
- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**Mobile Environment**
Please include:
- Device: iPhone 11 simulator
- iOS version: iOS 13.5
- WooCommerce iOS version: 4.8

**WordPress Environment**
<details>
```
Copy and paste the system status report from **WooCommerce > System Status** in WordPress admin.
```
</details>

### PR DESCRIPTION
**Pull request by jaclync**
Fixes #1985 
Fixes #2712

## Background about the changes

Shipping class is the trickiest field in shipping settings. Before this PR, we're passing the full `ProductShippingClass` from shipping settings to update a product. However, we don't always have `ProductShippingClass` around. A product/variation only has shipping class ID & slug fields, and currently we use the shipping class ID to sync the shipping class with the API in shipping settings VC's `viewDidLoad` so that we can display the name and set it as the selected shipping class in the shipping class list selector. In this PR, I replaced the passing of `ProductShippingClass` with shipping class ID and shipping class slug. We need both the slug and ID because:
- The full shipping class is fetched from the API given the shipping class ID
- The slug is used to determine if a product has a shipping class (empty or nil), since a shipping class ID is not optional and it's not very clear how to interpret a shipping class ID `0`

When testing this PR, I also noticed a pre-existing issue: https://github.com/woocommerce/woocommerce-ios/issues/2712 where a shipping class is removed after tapping "Done" before the shipping class is fetched because the use of `ProductShippingClass` (it's nil until it's synced). This PR fixes this issue as well.

## Changes

- In `ProductFormViewModelProtocol`, replaced the shipping class `shippingClass: ProductShippingClass` in `updateShippingSettings` with `shippingClass: String?` and `shippingClassID: Int64?`. Updated its implementations and affected unit tests
- Moved view data from `ProductShippingSettingsViewController` to `ProductShippingSettingsViewModel` (that conforms to 2 new protocols `ProductShippingSettingsViewModelOutput` and `ProductShippingSettingsActionHandler`). Updated `ProductFormViewController` with related changes, and added unit tests for edge cases
- Reorganized the unit tests on product editing a bit by creating folders like "Edit inventory", "Edit shipping", and "Edit price"

## Testing

No user-facing changes are expected, except for the fix for #2712. Please sanity check on the inventory settings for a product and product variation:

### Product shipping

- Go to the Products tab
- Tap on a physical product
- Tap on the shipping row --> the shipping settings should work like before
- Try updating the shipping class and maybe other field(s) then tap "Done"
- Tap "Update" --> the product should be updated remotely

### Product variation shipping

- Go to the Products tab
- Tap on a variable product
- Tap on the variations row
- Tap on a variation
- Tap on the shipping row --> there should be no "shipping class" section, and the rest of shipping settings should work like before
- Try updating some field(s) and tap "Done"
- Tap "Update" --> the variation should be updated remotely

### Bug fix #2712

- Go to the Products tab
- Tap on a physical product that has a shipping class set
- Tap on the shipping row and quickly tap "Done" before the shipping class row shows the name of the shipping class --> there should not be a "Update" CTA in the navigation bar, and the product's shipping class remains the same

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Products/Edit Product/Shipping Settings/ProductShippingSettingsViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    private var originalShippingClass: ProductShippingClass?
+
+    init(product: ProductFormDataModel) {
+        self.product = product
+        self.weight = product.weight
+        self.length = product.dimensions.length
+        self.width = product.dimensions.width
+        self.height = product.dimensions.height
+        self.shippingClassSlug = product.shippingClass
+        self.shippingClassID = product.shippingClassID
```
Comment (pmusolino) [inline WooCommerce/Classes/ViewRelated/Products/Edit Product/Shipping Settings/ProductShippingSettingsViewModel.swift]: nit: `self` is not needed 👍

File: WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/Edit Shipping/ProductShippingSettingsViewModelTests.swift [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        XCTAssertEqual(viewModel.weight, "1.6")
+        XCTAssertEqual(viewModel.length, dimensions.length)
+        XCTAssertEqual(viewModel.width, dimensions.width)
+        XCTAssertEqual(viewModel.height, dimensions.height)
+        XCTAssertNil(viewModel.shippingClass)
+    }
+
+    // MARK: - `completeUpdating`
+
+    func test_shipping_values_remain_the_same_after_completing_update_before_shipping_class_API_sync() {
```
Comment (pmusolino) [inline WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/Edit Shipping/ProductShippingSettingsViewModelTests.swift:39]: 👏

File: WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/Edit Shipping/ProductShippingSettingsViewModelTests.swift [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+                XCTAssertEqual(shippingClassID, 0)
+                XCTAssertTrue(hasUnsavedChanges)
+                expectation.fulfill()
+            }
+        }
+    }
+
+    // MARK: `hasUnsavedChanges`
+
+    func test_shipping_class_API_sync_with_a_different_slug_results_in_no_unsaved_changes() {
```
Comment (pmusolino) [inline WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/Edit Shipping/ProductShippingSettingsViewModelTests.swift]: nit:
```suggestion
    func test_shipping_class_API_sync_with_a_different_slug_has_no_unsaved_changes() {
```

File: WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/Edit Shipping/ProductShippingSettingsViewModelTests.swift [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        let viewModel = ProductShippingSettingsViewModel(product: model)
+        let retrievedShippingClass = ProductShippingClass(count: 0, descriptionHTML: nil, name: "60 Days", shippingClassID: 2, siteID: 0, slug: "90-day")
+        viewModel.onShippingClassRetrieved(shippingClass: retrievedShippingClass)
+        let hasUnsavedChanges = viewModel.hasUnsavedChanges()
+
+        // Assert
+        XCTAssertFalse(hasUnsavedChanges)
+    }
+
+    func test_updating_with_the_same_values_results_in_no_unsaved_changes() {
```
Comment (pmusolino) [inline WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/Edit Shipping/ProductShippingSettingsViewModelTests.swift]: nit:
```suggestion
    func test_updating_with_the_same_values_has_no_unsaved_changes() {
```

File: WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/Edit Shipping/ProductShippingSettingsViewModelTests.swift [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+        let retrievedShippingClass = ProductShippingClass(count: 0, descriptionHTML: nil, name: "60 Days", shippingClassID: 2, siteID: 0, slug: "90-day")
+        viewModel.onShippingClassRetrieved(shippingClass: retrievedShippingClass)
+        viewModel.handleShippingClassChange(retrievedShippingClass)
+        let hasUnsavedChanges = viewModel.hasUnsavedChanges()
+
+        // Assert
+        XCTAssertFalse(hasUnsavedChanges)
+    }
+
+    func test_updating_with_different_values_results_in_unsaved_changes() {
```
Comment (pmusolino) [inline WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/Edit Shipping/ProductShippingSettingsViewModelTests.swift]: nit:
```suggestion
    func test_updating_with_different_values_has_unsaved_changes() {
```

**Review by pmusolino — APPROVED** [review]
Thank you Jaclyn for taking care of this! 💯 
It looks really good to me 🚢 

Code ✅ just some minor nits.
Testing ✅ everything works perfectly. I was also able to verify issue #2712 that I was able to replicate in `develop` and not here.

File: WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/Edit Shipping/ProductShippingSettingsViewModelTests.swift (hunk 5 above)
Comment (jaclync) [reply WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/Edit Shipping/ProductShippingSettingsViewModelTests.swift]: haha it was quite wordy before!

**Comment by jaclync** [comment]
Thanks for testing and the suggestions Paolo, I made the updates in https://github.com/woocommerce/woocommerce-ios/pull/2714/commits/653252c5e15e9a16a700ecf083be76d697fa589b 😄
